test/test_linalg.py

```diff
@@ -3701,23 +3701,30 @@ def test_qr_vs_numpy(self, device, dtype):
     @skipCUDAIfNoCusolver
     @skipCPUIfNoLapack
     @dtypes(torch.float)
-    def test_linalg_qr_autograd_errors(self, device, dtype):
-        # torch.linalg.qr(mode='r') returns only 'r' and discards 'q', but
-        # without 'q' you cannot compute the backward pass. Check that
-        # linalg_qr_backward complains cleanly in that case.
-        inp = torch.randn((5, 7), device=device, dtype=dtype, requires_grad=True)
-        q, r = torch.linalg.qr(inp, mode='r')
-        self.assertEqual(q.shape, (0,))  # empty tensor
-        b = torch.sum(r)
-        with self.assertRaisesRegex(RuntimeError,
-                                    "The derivative of linalg.qr depends on Q"):
-            b.backward()
-        inp = torch.randn((7, 5), device=device, dtype=dtype, requires_grad=True)
-        q, r = torch.linalg.qr(inp, mode='complete')
-        b = torch.sum(r)
-        with self.assertRaisesRegex(RuntimeError,
-                                    "The QR decomposition is not differentiable when mode='complete' and nrows > ncols"):
-            b.backward()
+    def test_linalg_qr_autograd(self, device, dtype):
+        # Check differentiability for modes as specified in the docs:
+        # Mode 'reduced' is always differentiable.
```

Comment:
I guess this is only true for full-rank inputs?

Reply:
@nikitaved good point. Should I try to work in both full-rank and non-full-rank inputs as testcases? 

I'm trying generally to match the docs (https://pytorch.org/docs/stable/generated/torch.linalg.qr.html) on testcases.
They're a little ambiguous on the behavior under full-rank inputs:
```
The QR decomposition is only well-defined if the first k = min(m, n) columns of every matrix in A are linearly independent. If this condition is not met, no error will be thrown, but the QR produced may be incorrect and its autodiff may fail or produce incorrect results.
```

It may also be easier on the reader to use the terminology of 'matrix rank' in the docs instead of if such and such columns are linearly independent.

Reply:
Let me try to add in testcases for differentiability based on the structure of the matrix.

Reply:
Will look at the implementation, but I think the docs are correct as concerns rank requirements for the QR operation to be both well-defined and differentiable: 

'
The QR decomposition is only well-defined if the first k = min(m, n) columns of every matrix in A are linearly independent. If this condition is not met, no error will be thrown, but the QR produced may be incorrect and its autodiff may fail or produce incorrect results.
'

which as I understand requires not a full rank matrix, but rather one of:

- a) A must be a tall matrix (m >= n) of full rank, or

- b) A must be a wide matrix (m < n) with the first k = min(m,n) columns linearly independent (which the end user may have to ensure beforehand with column pivoting).


I can restrict the matrix test case inputs to these forms a), b). As concerns behavior on incorrect rank/linear independence matrix inputs, however, would it be better for the `torch.linalg.qr` implementation to raise an error / give a warning? (current behavior from the docs- 'If this condition is not met, no error will be thrown, but the QR produced may be incorrect and its autodiff may fail or produce incorrect results.') It's a little vague as is.

Reply:
So, yeah, the input has to be full-rank. The burden is up to the user because checking this condition (i.e. if full-rank) is as hard as the decomposition itself, plus you need to sync devices to be able to throw an error depending on the value of the rank.

Reply:
What I am trying to say is that you generate inputs that are full-rank with high probability. I would just change the comment from "always differentiable" to "differentiable when full-rank".

Reply:
@nikitaved thanks for the explanation. I added comment lines on input rank requirements, and the random matrix inputs of testcases satisfying those requirements.